File: php/test/ProductEntityTest.php
<?php
declare(strict_types=1);

// Product entity test

require_once __DIR__ . '/../openfoodfacts_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class ProductEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = OpenfoodfactsSDK::test(null, null);
        $ent = $testsdk->Product(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = product_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "product." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set OPENFOODFACTS_TEST_PRODUCT_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $product_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.product")));
        $product_ref01_data = null;
        if (count($product_ref01_data_raw) > 0) {
            $product_ref01_data = Helpers::to_map($product_ref01_data_raw[0][1]);
        }

        // LOAD
        $product_ref01_ent = $client->Product(null);
        $product_ref01_match_dt0 = [];
        [$product_ref01_data_dt0_loaded, $err] = $product_ref01_ent->load($product_ref01_match_dt0, null);
        $this->assertNull($err);
        $this->assertNotNull($product_ref01_data_dt0_loaded);

    }
}

function product_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/product/ProductTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = OpenfoodfactsSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["product01", "product02", "product03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("OPENFOODFACTS_TEST_PRODUCT_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "OPENFOODFACTS_TEST_PRODUCT_ENTID" => $idmap,
        "OPENFOODFACTS_TEST_LIVE" => "FALSE",
        "OPENFOODFACTS_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["OPENFOODFACTS_TEST_PRODUCT_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["OPENFOODFACTS_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new OpenfoodfactsSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["OPENFOODFACTS_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["OPENFOODFACTS_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: rb/test/product_entity_test.rb
# Product entity test

require "minitest/autorun"
require "json"
require_relative "../Openfoodfacts_sdk"
require_relative "runner"

class ProductEntityTest < Minitest::Test
  def test_create_instance
    testsdk = OpenfoodfactsSDK.test(nil, nil)
    ent = testsdk.Product(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "product" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = OpenfoodfactsSDK.test(seed, nil)
    seen = base.Product(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = OpenfoodfactsConfig.shared_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = OpenfoodfactsSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.Product(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = product_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "product." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set OPENFOODFACTS_TEST_PRODUCT_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    product_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.product")))
    product_ref01_data = nil
    if product_ref01_data_raw.length > 0
      product_ref01_data = Helpers.to_map(product_ref01_data_raw[0][1])
    end

    # LIST
    product_ref01_ent = client.Product(nil)
    product_ref01_match = {}

    product_ref01_list_result = product_ref01_ent.list(product_ref01_match, nil)
    assert product_ref01_list_result.is_a?(Array)

    # LOAD
    product_ref01_match_dt0 = {
      "id" => product_ref01_data["id"],
    }
    product_ref01_data_dt0_loaded = product_ref01_ent.load(product_ref01_match_dt0, nil)
    product_ref01_data_dt0_load_result = Helpers.to_map(product_ref01_data_dt0_loaded.respond_to?(:data_get) ? product_ref01_data_dt0_loaded.data_get : product_ref01_data_dt0_loaded)
    assert !product_ref01_data_dt0_load_result.nil?
    assert_equal product_ref01_data_dt0_load_result["id"], product_ref01_data["id"]

  end
end

def product_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "product", "ProductTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = OpenfoodfactsSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["product01", "product02", "product03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["OPENFOODFACTS_TEST_PRODUCT_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "OPENFOODFACTS_TEST_PRODUCT_ENTID" => idmap,
    "OPENFOODFACTS_TEST_LIVE" => "FALSE",
    "OPENFOODFACTS_TEST_EXPLAIN" => "FALSE",
    "OPENFOODFACTS_APIKEY" => "",
  })

  idmap_resolved = Helpers.to_map(
    env["OPENFOODFACTS_TEST_PRODUCT_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["OPENFOODFACTS_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      # FIRST, so the generated fields below win: sdk-test-control.json's
      # test.client.options adds to the live client, it does not redirect it.
      Runner.live_client_options,
      {
        "apikey" => env["OPENFOODFACTS_APIKEY"],
      },
      extra || {},
    ])
    client = OpenfoodfactsSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["OPENFOODFACTS_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["OPENFOODFACTS_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
